### PR TITLE
Improvements on neighbors-based estimators

### PR DESCRIPTION
**Pull request by jacobmontiel**
Changes proposed in this pull request:

- Create new base class `BaseNeighbors` for neighbors-based estimators
- Add implementation of `KNNRegressor`
- Remove in-house `KDTree` implementation in favor of the `sklearn` implementation which is optimized
- Replace `InstanceWindow` by simpler `SlidingWindow` method
- Improve documentation of `KNNClassifier` and `KNNADWINClassifier`
- Update related tests
- Add new tests for KNN regressor

---

#### Implementation
- [x] Implementation is correct (it performs its intended function).
- [x] Code complies with PEP-8 and is consistent with the framework.
- [x] Code is properly documented.
- [x] PR description covers ALL the changes performed.
- [x] Files changed (update, add, delete) are in the PR's scope (no extra files are included).

#### Tests
- [x] New functionality is tested.
- [x] Tests are created for the new functionality or existing tests are updated accordingly.
- [x] ALL tests pass with no errors.
- [x] CI/CD pipelines run with no errors.
- [x] Test Coverage is maintained (coverage may drop by no more than 0.2%).

### CODE REVIEW

File: src/skmultiflow/lazy/base_neighbors.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        self.leaf_size = leaf_size
+        if metric not in self.valid_metrics():
+            raise ValueError("Invalid metric: {}.\n"
+                             "Valid options are: {}".format(metric,
+                                                            self.valid_metrics()))
+        self.metric = metric
+        self.data_window = SlidingWindow(window_size=max_window_size)
+
+    def _get_neighbors(self, X):
+        tree = KDTree(self.data_window.features_buffer, self.leaf_size, metric=self.metric, )
```
Comment (smastelini) [inline src/skmultiflow/lazy/base_neighbors.py:26]: Is there a reason for the comma at the end of KDTree's construction?

File: src/skmultiflow/lazy/knn_adwin.py [hunk 2]
```diff
@@ -45,30 +45,26 @@ class KNNADWINClassifier(KNNClassifier):
...
-    before at least k samples have been analyzed by the algorithm.
+
+    metric: string or sklearn.DistanceMetric object
+        sklearn.KDTree parameter. The distance metric to use for the KDTree.
+        Default=’euclidean’. KNNClassifier.valid_metrics() gives a list of
+        the metrics which are valid for KDTree.
+
+    Notes
+    -----
+    This estimator is not optimal for a mixture of categorical and numerical
```
Comment (smastelini) [inline src/skmultiflow/lazy/knn_adwin.py:56]: Is there a workaround for this case? Should we explicit that categoric inputs are going to be treated as numeric features?

File: src/skmultiflow/lazy/knn_classifier.py [hunk 3]
```diff
@@ -1,49 +1,54 @@
...
+
+    metric: string or sklearn.DistanceMetric object
+        sklearn.KDTree parameter. The distance metric to use for the KDTree.
+        Default=’euclidean’. KNNClassifier.valid_metrics() gives a list of
+        the metrics which are valid for KDTree.
     
     Notes
     -----
-    This classifier is not optimal for a mixture of categorical and numerical features.
+    This estimator is not optimal for a mixture of categorical and numerical
```
Comment (smastelini) [inline src/skmultiflow/lazy/knn_classifier.py:50]: Should we explicit that categorical features are going to be treated as numeric?

File: src/skmultiflow/lazy/knn_classifier.py [hunk 4]
```diff
@@ -164,49 +158,26 @@ def predict_proba(self, X):
         Returns
         -------
         numpy.ndarray
-            An array of shape (n_samples, n_features), in which each outer entry is 
-            associated with the X entry of the same index. And where the list in 
-            index [i] contains len(self.target_value) elements, each of which represents
-            the probability that the i-th sample of X belongs to a certain label.
+            A  2D array of shape (n_samples, n_classes). Where each i-row
```
Comment (smastelini) [inline src/skmultiflow/lazy/knn_classifier.py:161]: What about "each i-th row"

File: src/skmultiflow/lazy/base_neighbors.py (hunk 1 above)
Comment (jacobmontiel) [reply src/skmultiflow/lazy/base_neighbors.py]: No. Removed in 31687ca4743f3fa03f752ed182135cbef9fc7901

File: src/skmultiflow/lazy/knn_adwin.py (hunk 2 above)
Comment (jacobmontiel) [reply src/skmultiflow/lazy/knn_adwin.py:56]: Good idea. Updated in 31687ca4743f3fa03f752ed182135cbef9fc7901

File: src/skmultiflow/lazy/knn_classifier.py (hunk 3 above)
Comment (jacobmontiel) [reply src/skmultiflow/lazy/knn_classifier.py:50]: Updated in 31687ca4743f3fa03f752ed182135cbef9fc7901

File: src/skmultiflow/lazy/knn_classifier.py (hunk 4 above)
Comment (jacobmontiel) [reply src/skmultiflow/lazy/knn_classifier.py]: Updated in 31687ca4743f3fa03f752ed182135cbef9fc7901

**Pull request merged**